contracts/governance/Vesting/OriginInvestorsClaim.sol

```diff
@@ -80,46 +68,83 @@ contract OriginInvestorsClaim is Ownable {
 	}
 
 	//in case we have unclaimed tokens
-	function ownerTtransferBalance(address toAddress) public onlyOwner {
-		SOVToken.transfer(toAddress, SOVToken.balanceOf(address(this)));
+	function authorizedBalanceWithdraw(address toAddress) public onlyAuthorized {
+		require(
+			SOVToken.transfer(toAddress, SOVToken.balanceOf(address(this))),
+			"OriginInvestorsClaim::authorizedTransferBalance: transfer failed"
+		);
 	}
 
-	function setInvestorsAmountsList(address[] memory investors, uint256[] memory claimAmounts) public onlyAuthorized {
-		require(investors.length == claimAmounts.length, "investors.length != claimAmounts.length");
+	/**
+	 *  @notice the contract should be approved or transferred necessary amount of SOV prior to calling the function
+	 *	@param allowanceFrom address that approved total amount for this contract or 0x0 if transferred
+	 *  @param investors is the list of investors addresses
+	 *  @param claimAmounts is the list of amounts for investors investors[i] will receive claimAmounts[i] of SOV
+	 */
+	function setInvestorsAmountsList(
+		address[] memory investors,
+		uint256[] memory claimAmounts,
+		address allowanceFrom
+	) public onlyAuthorized {
+		uint256 totalAmount;
+		require(
+			investors.length == claimAmounts.length,
+			"OriginInvestorsClaim::setInvestorsAmountsList: investors.length != claimAmounts.length"
+		);
+		require(!investorsListInitialized, "OriginInvestorsClaim::setInvestorsAmountsList: the investors list has already been set");
+		investorsListInitialized = true;
+
 		for (uint256 i = 0; i < investors.length; i++) {
 			investorsAmountsList[investors[i]] = claimAmounts[i];
+			totalAmount = totalAmount.add(claimAmounts[i]);
 		}
 
-		emit InvestorsAmountsListSet(investors.length);
+		require(
+			SOVToken.balanceOf(address(this)) >= totalAmount ||
+				(allowanceFrom != address(0) && SOVToken.transferFrom(allowanceFrom, address(this), totalAmount)),
+			"OriginInvestorsClaim::setInvestorsAmountsList: the contract is not enough financed or wrong allowance address passed"
+		);
+
+		emit InvestorsAmountsListSet(investors.length, totalAmount);
 	}
 
 	function claim() public onlyWhitelisted {
 		if (now < vestingTerm) {
 			createVesting();
 		} else {
-			withdraw();
+			transfer();
 		}
 	}
 
-	function createVesting() internal notInVestingList {
+	function createVesting() internal {
 		uint256 cliff = vestingTerm.sub(now);
 		uint256 duration = cliff;
 		uint256 amount = investorsAmountsList[msg.sender];
 		address vestingContractAddress;
 
-		vestingList[msg.sender] = true;
+		vestingContractAddress = vestingRegistry.getVesting(msg.sender);
+		require(vestingContractAddress == address(0), "OriginInvestorsClaim::withdraw: investor has an active vesting contract");
```

Comment:
are you sure that an investor cannot have another vesting already? from "some other source"

Reply:
excellent point. we need to consider that a lot of origin investors already invested in genesis. so, that will likely be the case for some. these people still need to be able to claim their tokens. therefore, they should be able to pass an address which should be the token owner.

Reply:
this also means that we need to communicate to our users that they HAVE to pass another address if they invested in genesis. victor needs to know that the frontend should check if the user is cSOV holder.

Reply:
I haven't dived deep into the vesting, however I have a feeling that it is possible to retrieve an existing vesting contract and add more funds there (i.e. reuse existing user's vesting contract for origin vesting). Initial code's logic was close to that behavior.